agent/engine/execcmdagent/manager_linux.go

```diff
@@ -0,0 +1,165 @@
+	if !task.IsExecCommandAgentEnabled() {
+		return NotRestarted, nil
+	}
+	execMD := container.GetExecCommandAgentMetadata()
+	if execMD == nil { // This means m.Start has never been invoked
+		return NotRestarted, nil
+	}
+	res, err := m.inspectExecAgentProcess(ctx, client, execMD)
+	if err != nil {
+		// We don't want to report InspectContainerExec errors, just that we don't know what the status of the agent is
+		return Unknown, nil
+	}
+	if res.Running { // agent still running, nothing to do
+		return NotRestarted, nil
+	}
+	// Restart if not running
+	//TODO: [ecs-exec] retry only for certain exit codes?
+	seelog.Warnf("Task engine [%s]: ExecCommandAgent Process stopped (exitCode:=%d) for container %s, restarting...", task.Arn, res.ExitCode, containerId)
+	container.SetExecCommandAgentMetadata(nil)
+	err = m.Start(ctx, client, task, container, containerId)
+	if err != nil {
+		return NotRestarted, err
+	}
+	return Restarted, nil
+}
+
+func (m *manager) inspectExecAgentProcess(ctx context.Context, client dockerapi.DockerClient, execMD *apicontainer.ExecCommandAgentMetadata) (*types.ContainerExecInspect, error) {
+	backoff := retry.NewExponentialBackoff(m.retryMinDelay, m.retryMaxDelay, retryJitterMultiplier, retryDelayMultiplier)
+	ctx, cancel := context.WithTimeout(ctx, m.inspectRetryTimeout)
+	defer cancel()
+	var (
+		inspectRes *types.ContainerExecInspect
+		inspectErr error
+	)
+	retry.RetryNWithBackoffCtx(ctx, backoff, maxRetries, func() error {
+		inspectRes, inspectErr = client.InspectContainerExec(ctx, execMD.DockerExecID, dockerclient.ContainerExecInspectTimeout)
+		if inspectErr != nil {
+			retryable := true
+			if _, ok := inspectErr.(*dockerapi.DockerTimeoutError); ok {
+				retryable = false
+			}
+			return StartError{
+				error:     inspectErr,
+				retryable: retryable,
+			}
+		}
+		return nil
+	})
+	return inspectRes, inspectErr
+}
+
+// Start idempotently starts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// In order for the ExecCommandAgent to be successfully started, the following conditions need to be met:
+//	1) There shall be at most one ExecCommandAgent process running in the container (with the name defined in apitask.ExecCommandAgentBinName)
+//	2) If an ExecCommandAgent process is already running in the container, its PID shall be equal to container.GetExecCommandAgentMetadata().PID
+// If any of the previous conditions is violated, ExecCommandAgent's can't be verified and an error is returned.
+// If no error is returned, it can be assumed the ExecCommandAgent is started.
+func (m *manager) Start(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) error {
+	if !task.IsExecCommandAgentEnabled() {
+		return nil
```

Comment:
debug log here?  why are we calling start on non execute-command enabled tasks?

Reply:
Will add in next rev